### PR TITLE
Fix volume backend caching

### PR DESCRIPTION
**Pull request by wdurairaj**
Some 16 ut failures are outstanding. Hence marking as WIP

### CODE REVIEW

**Comment by imran-ansari** [comment]
Hi William,
I think for import volume, we have introduced extra get_vol_by_name call. We will need to fix that. That will fix some.
For others, let me have a look.

Regards,
Imran

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by hpe-storage** [comment]
Hi Imran,


I have tested the change extensively with 15 parallel mounts for volume in RCG, plus some other tests like importvol, snapshot,clone.


So, it's only the UT which is haunting me.


Thanks & Regards,

William.

________________________________

**Comment by imran-ansari** [comment]
Looking into it.

**Comment by wdurairaj** [comment]
Hi  Imran,

Doing this change reduces the UT failures to 2, however, we will not do this based on your earlier comment.
```
docker@csosbe03b10:~/python-hpedockerplugin$ git diff
diff --git a/hpedockerplugin/backend_orchestrator.py b/hpedockerplugin/backend_orchestrator.py
index 8dd0168..8d1b895 100644
--- a/hpedockerplugin/backend_orchestrator.py
+++ b/hpedockerplugin/backend_orchestrator.py
@@ -115,7 +115,8 @@ class Orchestrator(object):
                 # where the backend can't be read from volume
                 # metadata in etcd
                 LOG.info(' vol obj read from etcd : %s' % vol)
-                raise exception.HPEPluginReadBackendFailed(volname=volname)
+                return 'DEFAULT'
+                #raise exception.HPEPluginReadBackendFailed(volname=volname)
         finally:
             self.volume_backend_lock.release()

@@ -147,7 +148,8 @@ class Orchestrator(object):
             LOG.debug('Removing entry for volume %s from cache' %
                       volname)
             # This if condition is to make the test code happy
-            if volname in self.volume_backends_map:
+            if volname in self.volume_backends_map and \
+               ret_val is not None:
                 del self.volume_backends_map[volname]
         return ret_val
```

**Comment by wdurairaj** [comment]
2 ut's are outstanding with the latest commit -- https://api.travis-ci.com/v3/job/159264761/log.txt

```
======================================================================
FAIL: test.test_hpe_plugin_v2.HpeDockerFCUnitTests.test_remove_regular_volume
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/test_hpe_plugin_v2.py", line 37, in banner_wrapper
    func(self, *args, **kwargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/test_hpe_plugin_v2.py", line 405, in test_remove_regular_volume
    test.run_test(self)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 119, in run_test
    self._mock_execute_api(plugin_api=self._get_plugin_api())
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/.tox/py34/lib/python3.4/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/setup_mock.py", line 74, in setup_mock_wrapper
    return func(self, mock_objects, *args, **kwargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 109, in _mock_execute_api
    raise ex
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 102, in _mock_execute_api
    self.check_response(resp)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/removevolume_tester.py", line 28, in check_response
    self._test_case)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/removevolume_tester.py", line 47, in check_response
    mock_3parclient.deleteVolume.assert_called()
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/.tox/py34/lib/python3.4/site-packages/mock/mock.py", line 906, in assert_called
    raise AssertionError(msg)
AssertionError: Expected 'deleteVolume' to have been called.
======================================================================
FAIL: test.test_hpe_plugin_v2.HpeDockerISCSIUnitTests.test_remove_regular_volume
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/test_hpe_plugin_v2.py", line 37, in banner_wrapper
    func(self, *args, **kwargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/test_hpe_plugin_v2.py", line 405, in test_remove_regular_volume
    test.run_test(self)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 119, in run_test
    self._mock_execute_api(plugin_api=self._get_plugin_api())
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/.tox/py34/lib/python3.4/site-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/setup_mock.py", line 74, in setup_mock_wrapper
    return func(self, mock_objects, *args, **kwargs)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 109, in _mock_execute_api
    raise ex
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/hpe_docker_unit_test.py", line 102, in _mock_execute_api
    self.check_response(resp)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/removevolume_tester.py", line 28, in check_response
    self._test_case)
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/test/removevolume_tester.py", line 47, in check_response
    mock_3parclient.deleteVolume.assert_called()
  File "/home/travis/build/hpe-storage/python-hpedockerplugin/.tox/py34/lib/python3.4/site-packages/mock/mock.py", line 906, in assert_called
    raise AssertionError(msg)
AssertionError: Expected 'deleteVolume' to have been called.

Ran 203 tests in 23.117s
FAILED (failures=2)
```

File: hpedockerplugin/backend_orchestrator.py [hunk 1]
```diff
@@ -88,19 +88,36 @@ def initialize_manager_objects(self, host_config, backend_configs):
...
-        if vol is not None and 'backend' in vol:
-            current_backend = vol['backend']
-
-        return current_backend
+        else:
+            return self.add_cache_entry(volname)
+
+    def add_cache_entry(self, volname):
+        # Using this style of locking
+        # https://docs.python.org/3/library/threading.htmls#using-locks-conditions-and-semaphores-in-the-with-statement
```
Comment (imran-ansari) [inline hpedockerplugin/backend_orchestrator.py]: Line going beyond 80 characters

Comment (imran-ansari) [inline hpedockerplugin/backend_orchestrator.py]: Line going beyond 80 characters

File: hpedockerplugin/backend_orchestrator.py [hunk 2]
```diff
@@ -88,19 +88,36 @@ def initialize_manager_objects(self, host_config, backend_configs):
...
+                # populate the volume backend map for caching
+                LOG.debug(' Populating cache %s, %s '
+                          % (volname, current_backend))
+                self.volume_backends_map[volname] = current_backend
+                return current_backend
+            else:
+                # throw an exception for the condition
+                # where the backend can't be read from volume
+                # metadata in etcd
+                LOG.info(' vol obj read from etcd : %s' % vol)
```
Comment (imran-ansari) [inline hpedockerplugin/backend_orchestrator.py]: Should this be:
"vol obj is either not found in ETCD or it doesn't have backend set"

**Comment by wdurairaj** [comment]
@imran-ansari  i will merge this change for now, since this changes are going for plugin_v2 and we can re-work on it further if there are any problems identified in CI regression run.

I'm merging this changes to get an fair idea of how these changes will affect volume operations. Once the CI build runs error free, we will address any review comments on this change (if any), and those changes + this PR changes will be merged to branch v300, and we will do a final RC build for Docker certification.

**Pull request merged**